core/src/avm2/globals/flash/utils/bytearray.rs

```diff
@@ -0,0 +1,929 @@
+    this: Option<Object<'gc>>,
+    _args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error> {
+    if let Some(mut bytearray) = this
+        .unwrap()
+        .as_bytearray_mut(activation.context.gc_context)
+    {
+        if let Ok(num) = bytearray.read_unsigned_int() {
+            return Ok(Value::Unsigned(num));
+        }
+    }
+    Ok(Value::Undefined)
+}
+
+pub fn read_boolean<'gc>(
+    activation: &mut Activation<'_, 'gc, '_>,
+    this: Option<Object<'gc>>,
+    _args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error> {
+    if let Some(mut bytearray) = this
+        .unwrap()
+        .as_bytearray_mut(activation.context.gc_context)
+    {
+        if let Ok(num) = bytearray.read_boolean() {
+            return Ok(Value::Bool(num));
+        }
+    }
+    Ok(Value::Undefined)
+}
+
+pub fn read_byte<'gc>(
+    activation: &mut Activation<'_, 'gc, '_>,
+    this: Option<Object<'gc>>,
+    _args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error> {
+    if let Some(mut bytearray) = this
+        .unwrap()
+        .as_bytearray_mut(activation.context.gc_context)
+    {
+        if let Ok(num) = bytearray.read_byte() {
+            return Ok(Value::Integer(num as i32));
+        }
+    }
+    Ok(Value::Undefined)
+}
+
+pub fn read_utf_bytes<'gc>(
+    activation: &mut Activation<'_, 'gc, '_>,
+    this: Option<Object<'gc>>,
+    args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error> {
+    if let Some(mut bytearray) = this
+        .unwrap()
+        .as_bytearray_mut(activation.context.gc_context)
+    {
+        if let Some(Value::Integer(len)) = args.get(0) {
+            if *len < 0 {
+                log::error!("ByteArray: Did not get proper length");
+                return Ok(Value::Undefined);
+            }
+            if let Ok(bytes) = bytearray.read_exactly(*len as usize) {
+                return Ok(AvmString::new(
+                    activation.context.gc_context,
+                    String::from_utf8_lossy(&bytes),
+                )
+                .into());
+            }
+        }
+    }
+    Ok(Value::Undefined)
+}
+
+pub fn read_unsigned_byte<'gc>(
+    activation: &mut Activation<'_, 'gc, '_>,
+    this: Option<Object<'gc>>,
+    _args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error> {
+    if let Some(mut bytearray) = this
+        .unwrap()
+        .as_bytearray_mut(activation.context.gc_context)
+    {
+        if let Ok(num) = bytearray.read_unsigned_byte() {
+            return Ok(Value::Unsigned(num as u32));
+        }
+    }
+    Ok(Value::Undefined)
+}
+
+pub fn write_float<'gc>(
+    activation: &mut Activation<'_, 'gc, '_>,
+    this: Option<Object<'gc>>,
+    args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error> {
+    if let Some(mut bytearray) = this
+        .unwrap()
+        .as_bytearray_mut(activation.context.gc_context)
+    {
+        if let Value::Number(num) = args.get(0).unwrap_or(&Value::Undefined) {
+            bytearray.write_float(*num as f32);
+        }
```

Comment:
I think that the argument should be converted to Number, instead of processing it only when it is Number.

My test code here.
```as
var test = new ByteArray();
test.writeFloat(null);
test.writeFloat({});
test.writeFloat("123");

test.position = 0;
trace(test.readFloat());
trace(test.readFloat());
trace(test.readFloat());
```

In Flash Player
```
0
NaN
123
```

In Ruffle
```
undefined
undefined
undefined
```